libs/utils/src/sync/heavier_once_cell.rs

```diff
@@ -69,37 +69,44 @@ impl<T> OnceCell<T> {
         F: FnOnce(InitPermit) -> Fut,
         Fut: std::future::Future<Output = Result<(T, InitPermit), E>>,
     {
-        let sem = {
-            let guard = self.inner.write().await;
-            if guard.value.is_some() {
-                return Ok(GuardMut(guard));
+        loop {
+            let sem = {
+                let guard = self.inner.write().await;
+                if guard.value.is_some() {
+                    return Ok(GuardMut(guard));
+                }
+                guard.init_semaphore.clone()
+            };
+
+            {
+                let permit = {
+                    // increment the count for the duration of queued
+                    let _guard = CountWaitingInitializers::start(self);
+                    sem.acquire().await
+                };
+
+                let Ok(permit) = permit else {
+                    let guard = self.inner.write().await;
+                    if !Arc::ptr_eq(&sem, &guard.init_semaphore) {
+                        // there was a take_and_deinit in between
+                        continue;
+                    }
+                    assert!(
+                        guard.value.is_some(),
+                        "semaphore got closed, must be initialized"
+                    );
+                    return Ok(GuardMut(guard));
+                };
+
+                permit.forget();
             }
-            guard.init_semaphore.clone()
-        };
 
-        let permit = {
-            // increment the count for the duration of queued
-            let _guard = CountWaitingInitializers::start(self);
-            sem.acquire_owned().await
-        };
-
-        match permit {
-            Ok(permit) => {
-                let permit = InitPermit(permit);
-                let (value, _permit) = factory(permit).await?;
+            let permit = InitPermit(sem);
+            let (value, _permit) = factory(permit).await?;
```

Comment:
On the common path of first-ever get_mut_or_init call to the cell, if the factory fails, we bail here and leave the semaphore in closed state, with `inner.value = None`


Ahh, no, `InitPermit` has a `Drop` impl. So, we rely on the factory to give the permit back to us.
Can't we just pass in `InitPermit` by reference to prevent that?

Also, I think a brief comment like

```
let permit = InitPermit(sem); // NB: re-ups semaphore on drop
```

wouldn't hurt.

Reply:
> Can't we just pass in InitPermit by reference to prevent that?

`Layer` needs to spawn a download task and send the token there, and get it back from the task.

> `let permit = InitPermit(sem); // NB: re-ups semaphore on drop`

There's `forget` in these functions which should be a clue, and the Drop impl.

Reply:
not clue enough for me but ok